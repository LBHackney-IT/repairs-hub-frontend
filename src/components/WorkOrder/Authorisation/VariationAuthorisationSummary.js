import PropTypes from 'prop-types'
import cx from 'classnames'
import {
  calculateCostBeforeVariation,
  calculateChangeInCost,
  calculateTotalVariedCost,
} from '../../../utils/helpers/calculations'
import { longDateToStr } from '../../../utils/date'

const VariationAuthorisationSummary = ({ variationTasks, originalSors }) => {
  const COST_BEFORE_VARIATION = 'Cost before variation'
  const TOTAL_VARIED_COST = 'Total cost after variation'
  const CHANGE_IN_COST = 'Change in cost'

  const sorStatus = (task) => {
    if (task.currentQuantity == 0 && task.originalQuantity == 0) {
      return 'New'
    }
    if (task.variedQuantity > task.currentQuantity) {
      return 'Increased'
    }
    if (task.variedQuantity < task.currentQuantity) {
      return 'Reduced'
    }
    return 'Unchanged'
  }
  const getCost = (task) => {
    return task.unitCost ? task.unitCost : 0
  }

  //cost calculation before variation, change in cost and total varied coast
  const costBeforeVariation = {
    description: COST_BEFORE_VARIATION,
    cost: calculateCostBeforeVariation(variationTasks.tasks),
  }
  const changeInCost = {
    description: CHANGE_IN_COST,
    cost: calculateChangeInCost(variationTasks.tasks),
  }
  const totalVaried = {
    description: TOTAL_VARIED_COST,
    cost: calculateTotalVariedCost(variationTasks.tasks),
  }

  //show cost breakdown
  const showCostBreakdown = () => {
    return [costBeforeVariation, changeInCost, totalVaried].map(
      (object, index) => (
        <tr
          key={index}
          className="govuk-table__row"
          id={object.description.toLowerCase().replace(/\s/g, '-')}
        >
          <th scope="row">{}</th>
          <th scope="row">{}</th>
          <th scope="row">{}</th>
          <th scope="row">{}</th>
          <th scope="row">{}</th>

          <td
            className={cx('govuk-!-padding-top-3', {
              'border-top-black': object.description !== COST_BEFORE_VARIATION,
            })}
          >
            <strong>{object.description}</strong>
          </td>
          <td
            className={cx('govuk-!-padding-top-3 govuk-table__cell--numeric', {
              'border-top-black': object.description !== COST_BEFORE_VARIATION,
            })}
          >
            £{object.cost}
          </td>
        </tr>
      )
    )
  }

  return (
    <>
      <h2 className="lbh-heading-h2">Summary of Tasks and SORs</h2>
      <h4 className="lbh-heading-h4">Original SORs</h4>
      <table className="govuk-table lbh-table original-sor-summary">
        <thead className="govuk-table__head">
          <tr className="govuk-table__row">
            <th scope="col" className="govuk-table__header">
              SOR code
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Quantity
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Cost
            </th>
          </tr>
        </thead>
        <tbody className="govuk-table__body">
          {originalSors.map((task, index) => (
            <tr index={index} key={index} className="govuk-table__row">
              <td className="govuk-table__cell">
                {task.code} - {task.description}
              </td>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                {task.originalQuantity}
              </td>
              <td className="govuk-table__cell govuk-table__cell--numeric">
                £{task.cost}
              </td>
            </tr>
          ))}
        </tbody>
      </table>

      <h4 className="lbh-heading-h4">Updated Tasks SORs</h4>
      <br></br>
      <p className="govuk-body">Updated by: {variationTasks.authorName} </p>
      <p className="govuk-body">
        {longDateToStr(variationTasks.variationDate)}{' '}
      </p>
      <div className="lbh-stat">
        <span className="lbh-stat__caption">{variationTasks.notes}</span>
      </div>

      <table className="govuk-table lbh-table updated-tasks-table">
        <thead className="govuk-table__head">
          <tr className="govuk-table__row">
            <th scope="col" className="govuk-table__header">
              SOR Status
            </th>
            <th scope="col" className="govuk-table__header">
              SOR code
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Unit Cost
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Quantity
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Cost
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Varied quantity
            </th>
            <th
              scope="col"
              className="govuk-table__header govuk-table__header--numeric"
            >
              Varied Cost
            </th>
          </tr>
        </thead>
        <tbody className="govuk-table__body">
          {variationTasks.tasks
            ? variationTasks.tasks.map((task, index) => (
                <tr index={index} key={index} className="govuk-table__row">
                  <td className="govuk-table__cell">
                    <span
                      className={`status status-${sorStatus(task)
                        .replace(/\s+/g, '-')
                        .toLowerCase()}`}
                    >
                      {sorStatus(task)}
                    </span>
                  </td>
                  <td className="govuk-table__cell">
                    {task.code}
                    <p>{task.description}</p>
                  </td>
                  <td className="govuk-table__cell govuk-table__cell--numeric">
                    £{getCost(task)}
                  </td>
                  <td className="govuk-table__cell govuk-table__cell--numeric">
                    {task.currentQuantity}
                  </td>
                  <td className="govuk-table__cell govuk-table__cell--numeric">
                    £{getCost(task) * task.currentQuantity}
                  </td>
                  <td className="govuk-table__cell govuk-table__cell--numeric">
                    {task.variedQuantity}
                  </td>
                  <td className="govuk-table__cell govuk-table__cell--numeric">
                    £{getCost(task) * task.variedQuantity}
                  </td>
                </tr>
              ))
            : ''}
          {variationTasks.tasks ? showCostBreakdown() : ''}
        </tbody>
      </table>
    </>
  )
}

VariationAuthorisationSummary.propTypes = {
  variationTasks: PropTypes.object.isRequired,
  originalSors: PropTypes.array.isRequired,
}

export default VariationAuthorisationSummary
