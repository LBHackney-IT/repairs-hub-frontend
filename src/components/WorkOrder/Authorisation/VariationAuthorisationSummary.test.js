import { render } from '@testing-library/react'
import VariationAuthorisationSummary from './VariationAuthorisationSummary'

describe('VariationAuthorisationSummary component', () => {
  const props = {
    variationTasks: {
      notes: 'Variation reason: More work needed',
      tasks: [
        {
          code: 'DES5R005',
          currentQuantity: 1,
          description: 'Normal Call outs',
          id: 'abcd',
          originalQuantity: 1,
          unitCost: 4,
          variedQuantity: 4000,
        },
        {
          code: 'DES5R006',
          currentQuantity: 10,
          description: 'Normal Call outs',
          id: 'abc',
          originalQuantity: 1,
          unitCost: 19,
          variedQuantity: 2,
        },
      ],
      authorName: 'John Johnson (Alphatrack)',
      variationDate: '2021-05-11T16:04:05.008Z',
    },

    originalSors: [
      {
        id: 'ade7c53b-8947-414c-b88f-9c5e3d875cbf',
        code: 'DES5R006',
        description: 'Urgent call outs',
        dateAdded: '2021-02-03T09:33:35.757339',
        quantity: 2,
        cost: 10,
        status: 'Unknown',
        original: true,
        originalQuantity: 1,
      },
    ],
  }
  it('should render properly', () => {
    const { asFragment } = render(
      <VariationAuthorisationSummary
        variationTasks={props.variationTasks}
        originalSors={props.originalSors}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
