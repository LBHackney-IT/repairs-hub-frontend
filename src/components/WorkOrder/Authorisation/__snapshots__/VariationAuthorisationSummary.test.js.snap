// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariationAuthorisationSummary component should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Summary of Tasks and SORs
  </h2>
  <section
    class="lbh-collapsible"
  >
    <div
      aria-expanded="true"
      class="lbh-collapsible__button original-sors"
    >
      <h3
        class="lbh-heading-h3 lbh-collapsible__heading"
      >
        Original SORs
      </h3>
      <svg
        height="10"
        viewBox="0 0 17 10"
        width="17"
      >
        <path
          d="M2 1.5L8.5 7.5L15 1.5"
          stroke-width="3"
        />
      </svg>
    </div>
    <div
      class="lbh-collapsible__content undefined"
    >
      <table
        class="govuk-table lbh-table lbh-table original-sor-summary"
      >
        <thead
          class="govuk-table__head"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="col"
            >
              SOR code
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Quantity
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Cost
            </th>
          </tr>
        </thead>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <td
              class="govuk-table__cell"
            >
              DES5R006 - Urgent call outs
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              1
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £10
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </section>
  <section
    class="lbh-collapsible"
  >
    <div
      aria-expanded="true"
      class="lbh-collapsible__button updated-sors"
    >
      <h3
        class="lbh-heading-h3 lbh-collapsible__heading"
      >
        Updated Tasks SORs
      </h3>
      <svg
        height="10"
        viewBox="0 0 17 10"
        width="17"
      >
        <path
          d="M2 1.5L8.5 7.5L15 1.5"
          stroke-width="3"
        />
      </svg>
    </div>
    <div
      class="lbh-collapsible__content undefined"
    >
      <p
        class="lbh-body"
      >
        Updated by: John Johnson (Alphatrack) 
      </p>
      <p
        class="lbh-body"
      >
        Tuesday 11 May 2021 
      </p>
      <div
        class="lbh-stat"
      >
        <span
          class="lbh-stat__caption"
        >
          Variation reason: More work needed
        </span>
      </div>
      <table
        class="govuk-table lbh-table lbh-table updated-tasks-table"
      >
        <thead
          class="govuk-table__head"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="col"
            >
              SOR Status
            </th>
            <th
              class="govuk-table__header"
              scope="col"
            >
              SOR code
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Unit Cost
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Quantity
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Cost
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Varied quantity
            </th>
            <th
              class="govuk-table__header govuk-table__header--numeric"
              scope="col"
            >
              Varied Cost
            </th>
          </tr>
        </thead>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
            data-row-id="0"
          >
            <td
              class="govuk-table__cell"
            >
              <span
                class="govuk-tag lbh-tag lbh-tag--status-increased"
              >
                Increased
              </span>
            </td>
            <td
              class="govuk-table__cell"
            >
              DES5R005
              <p>
                Normal Call outs
              </p>
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £4
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              1
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £4
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              4000
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £16000
            </td>
          </tr>
          <tr
            class="govuk-table__row"
            data-row-id="1"
          >
            <td
              class="govuk-table__cell"
            >
              <span
                class="govuk-tag lbh-tag lbh-tag--status-reduced"
              >
                Reduced
              </span>
            </td>
            <td
              class="govuk-table__cell"
            >
              DES5R006
              <p>
                Normal Call outs
              </p>
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £19
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              10
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £190
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              2
            </td>
            <td
              class="govuk-table__cell govuk-table__cell--numeric"
            >
              £38
            </td>
          </tr>
        </tbody>
      </table>
      <table
        class="govuk-table lbh-table lbh-table calculated-cost govuk-!-margin-top-0"
      >
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
            id="cost-before-variation"
          >
            <th
              class="govuk-table__header govuk-table__header border-none"
              scope="row"
            >
              <p
                class="govuk-body govuk-!-margin-bottom-0 lbh-!-font-weight-bold"
              >
                Budget code – Subjective:
                <br />
                H2555 - 200108 Gutter Clearance
              </p>
            </th>
            <td
              class="govuk-table__cell govuk-!-padding-top-3"
            >
              <strong>
                Cost before variation
              </strong>
            </td>
            <td
              class="govuk-table__cell govuk-!-padding-top-3 govuk-table__cell--numeric"
            >
              £194.00
            </td>
          </tr>
          <tr
            class="govuk-table__row"
            id="change-in-cost"
          >
            <th
              class="govuk-table__header border-none"
              scope="row"
            />
            <td
              class="govuk-table__cell govuk-!-padding-top-3 border-top-black"
            >
              <strong>
                Change in cost
              </strong>
            </td>
            <td
              class="govuk-table__cell govuk-!-padding-top-3 govuk-table__cell--numeric border-top-black"
            >
              £15844.00
            </td>
          </tr>
          <tr
            class="govuk-table__row"
            id="total-cost-after-variation"
          >
            <th
              class="govuk-table__header border-none"
              scope="row"
            />
            <td
              class="govuk-table__cell govuk-!-padding-top-3 border-top-black"
            >
              <strong>
                Total cost after variation
              </strong>
            </td>
            <td
              class="govuk-table__cell govuk-!-padding-top-3 govuk-table__cell--numeric border-top-black"
            >
              £16038.00
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </section>
</DocumentFragment>
`;
